### ISSUE TITLE
Carthage fetches (sub-)dependencies multiple times

### ISSUE THREAD

**Issue by ffried** [Carthage/Carthage#1432]
* `carthage version`: 0.17.2
* `xcodebuild -version`: Xcode 7.3.1, Build version 7D1014
* Are you using `--no-build`? Yes
* Are you using `--no-use-binaries`? Yes
* Are you using `--use-submodules`? No

We have a project with framework dependencies on which the core project depends. The frameworks depend on each other as well.

Example: 

**Cartfile Core**

```
git "https://somewhere.internal/git/user/Utilities" "develop"
git "https://somewhere.internal/git/user/Processing" "develop"
git "https://somewhere.internal/git/user/DataStorage" "develop"
```

**Cartfile Processing**

```
git "https://somewhere.internal/git/user/Utilities" "develop"
```

**Cartfile DataStorage**

```
git "https://somewhere.internal/git/user/Utilities" "develop"
git "https://somewhere.internal/git/user/Processing" "develop"
```

When running `carthage update --no-use-binaries --no-build` on the core project, it will now fetch "Utilities" three times and "Processing" twice.
Fetching them once should be enough for knowing if there are updates. Also for checking conflicting dependencies, the location (branch, commit hash, tag, ...) can be used. If the specified location differs, fetching the distinct locations should be enough to determine if there's a conflict.

Or am I missing something here?

**Comment by BobElDevil** [comment]
I have an open PR that attempts to alleviate this problem #1429. Basically it's because we're fetching on demand and stateless, we refetch every time if the dependency descriptor is 'fuzzy' in any way, such as a branch name.

**Comment by ffried** [comment]
Solved by #1429
